va/caa.go

```diff
@@ -116,7 +117,61 @@ func parseResults(results []caaResult) (*CAASet, error) {
 	return nil, nil
 }
 
-func (va *ValidationAuthorityImpl) parallelCAALookup(ctx context.Context, name string, lookuper func(context.Context, string) ([]*dns.CAA, error)) []caaResult {
+// If the fqdn has a direct alias among the list of CNAMEs, return it; otherwise
+// return "". Note that RFC 6844 only asks for tree-climbing on direct aliases,
+// not recursive aliases.
+func findAlias(source string, cnames []*dns.CNAME) string {
+	for _, cname := range cnames {
+		if cname.Header().Name == source {
+			return cname.Target
+		}
+	}
+	return ""
+}
+
+func parent(fqdn string) string {
+	labels := strings.Split(fqdn, ".")
+	if len(labels) > 1 {
+		return strings.Join(labels[1:], ".")
+	} else {
+		return ""
+	}
+}
+
+// Implement pre-erratum 5065 style tree-climbing CAA. Note: a strict
+// interpretation of pre-5065 indicates a linear lookup path - if there is any
+// CNAME at all, that precludes further tree-climbing on the original FQDN. This
+// is clearly wrong. We implement a hybrid approach that is strictly more
+// conservative: We always do full tree-climbing on the original FQDN (by virtue
+// of parallelCAALookup. When the LegacyCAA flag is enabled, we also
+// do linear tree climbing on single-level aliases.
+func (va *ValidationAuthorityImpl) treeClimbingLookupCAA(ctx context.Context, fqdn string) ([]*dns.CAA, []*dns.CNAME, error) {
+	target := fqdn
+	// Limit CNAME chasing to break CNAME loops
+	for i := 0; i < 8 && target != ""; i++ {
+		caas, cnames, err := va.dnsClient.LookupCAA(ctx, target)
+		if err != nil {
+			return nil, nil, err
+		} else if len(caas) > 0 {
+			return caas, nil, nil
+		} else if len(cnames) > 0 {
+			target = findAlias(target, cnames)
+		} else if i == 0 {
+			// Special case: If there are no CNAMEs or CAAs on the very first lookup,
+			// skip the tree-climbing, since that's already implemented by parallelCAALookup.
+			// This minimizes duplicate lookups.
+			return nil, nil, nil
+		} else {
+			// If we're not on the first iteration (i.e. we've found a CNAME), do tree climbing.
+			target = parent(target)
+		}
+	}
+	return nil, nil, nil
+}
+
+type lookuperFunc func(context.Context, string) ([]*dns.CAA, []*dns.CNAME, error)
```

Comment:
Can you add a comment on the `lookuperFunc`?

Reply:
Does this function type need a `[]*dns.CNAME` return? It seems like only `parallelCAALookup` calls a `lookuperFunc` and it throws away the CNAMEs returned.

Reply:
Good point. Removed the CNAME return from `lookuperFunc`.